Mathlib/Algebra/Module/Basic.lean

```diff
@@ -0,0 +1,779 @@
+
+set_option linter.deprecated false in
+@[deprecated] theorem two_smul' : (2 : R) • x = bit0 x :=
+  two_smul R x
+#align two_smul' two_smul'
+
+@[simp]
+theorem inv_of_two_smul_add_inv_of_two_smul [Invertible (2 : R)] (x : M) :
+    (⅟ 2 : R) • x + (⅟ 2 : R) • x = x :=
+  Convex.combo_self invOf_two_add_invOf_two _
+#align inv_of_two_smul_add_inv_of_two_smul inv_of_two_smul_add_inv_of_two_smul
+
+/-- Pullback a `Module` structure along an injective additive monoid homomorphism.
+See note [reducible non-instances]. -/
+@[reducible]
+protected def Function.Injective.module [AddCommMonoid M₂] [SMul R M₂] (f : M₂ →+ M)
+    (hf : Injective f) (smul : ∀ (c : R) (x), f (c • x) = c • f x) : Module R M₂ :=
+  { hf.distribMulAction f smul with
+    smul := (· • ·)
+    add_smul := fun c₁ c₂ x => hf <| by simp only [smul, f.map_add, add_smul]
+    zero_smul := fun x => hf <| by simp only [smul, zero_smul, f.map_zero] }
+#align function.injective.module Function.Injective.module
+
+/-- Pushforward a `Module` structure along a surjective additive monoid homomorphism. -/
+protected def Function.Surjective.module [AddCommMonoid M₂] [SMul R M₂] (f : M →+ M₂)
+    (hf : Surjective f) (smul : ∀ (c : R) (x), f (c • x) = c • f x) : Module R M₂ :=
+  { hf.distribMulAction f smul with
+    smul := (· • ·)
+    add_smul := fun c₁ c₂ x => by
+      rcases hf x with ⟨x, rfl⟩
+      simp only [add_smul, ← smul, ← f.map_add]
+    zero_smul := fun x => by
+      rcases hf x with ⟨x, rfl⟩
+      rw [← f.map_zero, ← smul, zero_smul] }
+#align function.surjective.module Function.Surjective.module
+
+/-- Push forward the action of `R` on `M` along a compatible surjective map `f : R →+* S`.
+
+See also `Function.Surjective.mulActionLeft` and `Function.Surjective.distribMulActionLeft`.
+-/
+@[reducible]
+def Function.Surjective.moduleLeft {R S M : Type _} [Semiring R] [AddCommMonoid M] [Module R M]
+    [Semiring S] [SMul S M] (f : R →+* S) (hf : Function.Surjective f)
+    (hsmul : ∀ (c) (x : M), f c • x = c • x) : Module S M :=
+  { hf.distribMulActionLeft f.toMonoidHom hsmul with
+    smul := (· • ·)
+    zero_smul := fun x => by rw [← f.map_zero, hsmul, zero_smul]
+    add_smul := hf.forall₂.mpr fun a b x => by simp only [← f.map_add, hsmul, add_smul] }
+#align function.surjective.module_left Function.Surjective.moduleLeft
+
+variable {R} (M)
+
+/-- Compose a `Module` with a `RingHom`, with action `f s • m`.
+
+See note [reducible non-instances]. -/
+@[reducible]
+def Module.compHom [Semiring S] (f : S →+* R) : Module S M :=
+  { MulActionWithZero.compHom M f.toMonoidWithZeroHom, DistribMulAction.compHom M (f : S →* R) with
+    smul := SMul.comp.smul f
+    -- Porting note: the `show f (r + s) • x = f r • x + f s • x ` wasn't needed in mathlib3.
+    -- Somehow, now that `SMul` is heterogeneous, it can't unfold earlier fields of a definition for
+    -- use in later fields.  See
+    -- https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/
```

Comment:
Does the linter complain if this is on a single line? I thought it allowed long URL, but I am not sure.

Reply:
Looks like it works -- thanks!